src/huggingface_hub/repocard.py

```diff
@@ -1,30 +1,288 @@
-    ModelIndex,
-    SingleMetric,
-    SingleResult,
-    SingleResultDataset,
-    SingleResultTask,
+from huggingface_hub.hf_api import upload_file
+from huggingface_hub.repocard_data import (
+    CardData,
+    EvalResult,
+    eval_results_to_model_index,
+    model_index_to_eval_results,
 )
 
 from .constants import REPOCARD_NAME
+from .utils.logging import get_logger
 
 
 # exact same regex as in the Hub server. Please keep in sync.
+TEMPLATE_MODELCARD_PATH = Path(__file__).parent / "templates" / "modelcard_template.md"
 REGEX_YAML_BLOCK = re.compile(r"---[\n\r]+([\S\s]*?)[\n\r]+---[\n\r]")
 
-UNIQUE_RESULT_FEATURES = ["dataset", "task"]
-UNIQUE_METRIC_FEATURES = ["name", "type"]
+logger = get_logger(__name__)
+
+
+class ModelCard:
+    def __init__(self, content: str):
+        """Initialize a RepoCard from string content. The content should be a
+        Markdown file with a YAML block at the beginning and a Markdown body.
+
+        Args:
+            content (`str`): The content of the Markdown file.
+
+        Raises:
+            ValueError: When the content of the repo card metadata is not found.
+            ValueError: When the content of the repo card metadata is not a dictionary.
+        """
+        self.content = content
+        match = REGEX_YAML_BLOCK.search(content)
+        if match:
+            # Metadata found in the YAML block
+            yaml_block = match.group(1)
+            self.text = content[match.end() :]
+            data_dict = yaml.safe_load(yaml_block)
+
+            # The YAML block's data should be a dictionary
+            if not isinstance(data_dict, dict):
+                raise ValueError("repo card metadata block should be a dict")
+        else:
+            # Model card without metadata... create empty metadata
+            logger.warning(
+                "Repo card metadata block was not found. Setting CardData to empty."
+            )
+            data_dict = {}
+            self.text = content
+
+        model_index = data_dict.pop("model-index", None)
+        if model_index:
+            try:
+                model_name, eval_results = model_index_to_eval_results(model_index)
+                data_dict["model_name"] = model_name
+                data_dict["eval_results"] = eval_results
+            except KeyError:
+                logger.warning(
+                    "Invalid model-index. Not loading eval results into CardData."
+                )
+
+        self.data = CardData(**data_dict)
+
+    def __str__(self):
+        return f"---\n{self.data.to_yaml()}\n---\n{self.text}"
+
+    def save(self, filepath: Union[Path, str]):
+        r"""Save a RepoCard to a file.
+
+        Args:
+            filepath (`Union[Path, str]`): Filepath to the markdown file to save.
+
+        Example:
+            >>> from huggingface_hub import RepoCard
+            >>> card = RepoCard("---\nlanguage: en\n---\n# This is a test repo card")
+            >>> card.save("/tmp/test.md")
```

Comment:
```suggestion
        Example:
            ```python
            >>> from huggingface_hub import RepoCard
            >>> card = RepoCard("---\nlanguage: en\n---\n# This is a test repo card")
            >>> card.save("/tmp/test.md")
            ```
```

Reply:
(please apply the same to all code examples)

Reply:
is it ok if I add a new line before the closing backticks? this way doctests still run as expected (even though they aren't being used in our CI)